src/System.Linq/src/System/Linq/Concatenate.cs

```diff
@@ -186,29 +442,17 @@ public override bool MoveNext()
                 return false;
             }
 
-            public TSource[] ToArray()
+            public virtual TSource[] ToArray()
             {
                 return EnumerableHelpers.ToArray(this);
             }
 
-            public List<TSource> ToList()
+            public virtual List<TSource> ToList()
             {
-                var list = new List<TSource>();
-                for (int i = 0; ; i++)
-                {
-                    IEnumerable<TSource> source = GetEnumerable(i);
-                    if (source == null)
-                    {
-                        break;
-                    }
-
-                    list.AddRange(source);
-                }
-
-                return list;
+                return PopulateList(new List<TSource>());
```

Comment:
Assuming you get rid of Count and just use GetCount, this method shouldn't need to be made virtual nor should PopulateList need to be separated out into its own method.  You can just call GetCount(onlyIfCheap: true), using the result if it's >= 0.